src/datasets/arrow_dataset.py

```diff
@@ -2181,6 +2182,58 @@ def generator():
         logger.info(f"Finished writing TFRecord to {filename}")
         self = None  # delete the dataset reference used by tf_dataset
 
+    def to_dict(self, batch_size: Optional[int] = None, batched: bool = False) -> Union[dict, Iterator[dict]]:
+        """Returns the dataset as a Python dict. Can also return a generator for large datasets.
+
+        Args:
+            batched (`bool`): Set to :obj:`True` to return a generator that yields the dataset as batches of
+                :param:`batch_size` rows. Defaults to :obj:`False` (returns the whole datasetas once)
+            bacth_size (`Optional[int]`): The size (number of rows) of the batches if :param:`batched` is `True`.
+                Defaults to :obj:`.config.DEFAULT_MAX_BATCH_SIZE`.
+
+        Returns:
+            `dict` or `Iterator[dict]`
+        """
+        if not batched:
+            return self._data.to_pydict()
```

Comment:
You will need to use query_table here as well to take the indices mapping into account (`indices=self._indices.column(0) if self._indices is not None else None`).

Feel free to use the key `slice(0, len(self))` to query the full table after applying the indices mapping.